packages/grid/_modules_/grid/hooks/virtualization/useVirtualRows.ts

```diff
@@ -28,10 +28,7 @@ import { useApiEventHandler } from '../root/useApiEventHandler';
 type UseVirtualRowsReturnType = Partial<RenderContextProps> | null;
 
 // TODO v5: replace with @material-ui/core/utils/useEnhancedEffect.
-const useEnhancedEffect =
-  typeof window !== 'undefined' && process.env.NODE_ENV !== 'test'
-    ? React.useLayoutEffect
-    : React.useEffect;
+const useEnhancedEffect = typeof window !== 'undefined' ? React.useLayoutEffect : React.useEffect;
```

Comment:
why do we need to use blocking useLayoutEffect when window is not available?

Reply:
> If you use server rendering, keep in mind that neither useLayoutEffect nor useEffect can run until the JavaScript is downloaded. This is why React warns when a server-rendered component contains useLayoutEffect. To fix this, either move that logic to useEffect (if it isn’t necessary for the first render), or delay showing that component until after the client renders (if the HTML looks broken until useLayoutEffect runs).

https://reactjs.org/docs/hooks-reference.html#uselayouteffect

Reply:
The objective of using useLayoutEffect is to avoid displaying an intermediary state with an empty grid (useLayoutEffect needs to be opt-out on the server-side).

Reply:
why not use it directly? Have you tried with just useEffect?

Reply:
useEffect would raise a warning: 

> If you use server rendering, keep in mind that neither useLayoutEffect nor useEffect can run until the JavaScript is downloaded. This is why React warns when a server-rendered component contains useLayoutEffect. To fix this, either move that logic to useEffect (if it isn’t necessary for the first render), or delay showing that component until after the client renders (if the HTML looks broken until useLayoutEffect runs).

https://reactjs.org/docs/hooks-reference.html#uselayouteffect